File: leetcode/golang/300_longest_increasing_subsequence.go
func lengthOfLIS(nums []int) int {
    var dp []int = make([]int, len(nums));
    res := 0;
    for i := 0; i < len(nums); i++ {
        left, right := 0, res
        for left < right {
            mid := (left + right) / 2
            if dp[mid] < nums[i] {
                left = mid + 1
            } else {
                right = mid
            }
        }
        dp[left] = nums[i]
        if right == res {
            res++
        }
    }
    return res 
}

File: leetcode/c++/300_longest_increasing_subsequence.cpp
// https://leetcode.cn/problems/longest-increasing-subsequence/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int>dp(n, 1);
        int res = 0;
        for (int i = 0; i < n; i ++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i])
                    dp[i] = max(dp[i], dp[j] + 1);
            }
            res = max(res, dp[i]);
        }
        return res;
    }

    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int>dp(n);
        int res = 0;
        for (int num : nums) {
            int i = 0, j = res;
            while (i < j) {
                int m = (i + j) / 2;
                if (dp[m] < num) i = m + 1;
                else j = m;
            }
            dp[i] = num;
            if (j == res) res++;
        }
        return res;
    }
};

// 时间复杂度分别为O(n^2)和O(nlogn)，空间复杂度为O(n)